tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs

```diff
@@ -17,144 +20,246 @@ namespace Datadog.Trace.Agent.DiscoveryService
+            string? configurationEndpoint = null;
+            string? debuggerEndpoint = null;
+            string? statsEndpoint = null;
+            string? dataStreamsMonitoringEndpoint = null;
+
+            if (discoveredEndpoints is { Length: > 0 })
             {
-                return;
-            }
+                foreach (var discoveredEndpoint in discoveredEndpoints)
+                {
+                    var endpoint = discoveredEndpoint?.Trim('/');
+                    if (endpoint is null)
+                    {
+                        continue;
+                    }
 
-            ConfigurationEndpoint = SupportedConfigurationEndpoints
-               .FirstOrDefault(
-                    supportedEndpoint => discoveredEndpoints.Any(
-                        endpoint => endpoint.Trim('/').Equals(supportedEndpoint, StringComparison.OrdinalIgnoreCase)));
+                    // effectively a switch, but case insensitive
+                    if (endpoint.Equals(SupportedDebuggerEndpoint, StringComparison.OrdinalIgnoreCase))
+                    {
+                        debuggerEndpoint = endpoint;
+                    }
+                    else if (endpoint.Equals(SupportedConfigurationEndpoint, StringComparison.OrdinalIgnoreCase))
+                    {
+                        configurationEndpoint = endpoint;
+                    }
+                    else if (endpoint.Equals(SupportedStatsEndpoint, StringComparison.OrdinalIgnoreCase))
+                    {
+                        statsEndpoint = endpoint;
+                    }
+                    else if (endpoint.Equals(SupportedDataStreamsEndpoint, StringComparison.OrdinalIgnoreCase))
+                    {
+                        dataStreamsMonitoringEndpoint = endpoint;
+                    }
+                }
+            }
 
-            DebuggerEndpoint = SupportedDebuggerEndpoints
-               .FirstOrDefault(
-                    supportedEndpoint => discoveredEndpoints.Any(
-                        endpoint => endpoint.Trim('/').Equals(supportedEndpoint, StringComparison.OrdinalIgnoreCase)));
+            var existingConfiguration = _configuration;
 
-            StatsEndpoint = SupportedStatsEndpoints
-               .FirstOrDefault(
-                    supportedEndpoint => discoveredEndpoints.Any(
-                        endpoint => endpoint.Trim('/').Equals(supportedEndpoint, StringComparison.OrdinalIgnoreCase)));
+            if (existingConfiguration is null
+             || agentVersion != existingConfiguration.AgentVersion
+             || clientDropP0 != existingConfiguration.ClientDropP0s
+             || configurationEndpoint != existingConfiguration.ConfigurationEndpoint
+             || debuggerEndpoint != existingConfiguration.DebuggerEndpoint
+             || statsEndpoint != existingConfiguration.StatsEndpoint
+             || dataStreamsMonitoringEndpoint != existingConfiguration.DataStreamsMonitoringEndpoint)
```

Comment:
Just a Nit. You can convert `AgentConfiguration` from class to record, or make it `IEquatable`.
In that case, you can slightly simplify code by something like that:
```
            var newConfig = new AgentConfiguration(configurationEndpoint, debuggerEndpoint, agentVersion, statsEndpoint, dataStreamsMonitoringEndpoint, clientDropP0);
            if (newConfig != _configuration)
            {
                NotifySubscribers(newConfig);
            }
```

I think it will be more readable, and will cover possible future changes of `AgentConfiguration`

Reply:
I don't think we have added support for records yet have we? We need to add the init support - we discussed it a while back, and noone was too worried about it, but @OmerRaviv ultimately removed it. This seems like a good reason to add it though as I agree, this is pretty fragile right now 🤔

Reply:
We don't? 😨 
We _do_ use them a lot for the `LiveDebugger` related classes (records 😅), here are some examples:

https://github.com/DataDog/dd-trace-dotnet/blob/8c07d02c942e37d413972957eb7bd1140d230efd/tracer/src/Datadog.Trace/Debugger/PInvoke/ProbeStatus.cs#L12

https://github.com/DataDog/dd-trace-dotnet/blob/8c07d02c942e37d413972957eb7bd1140d230efd/tracer/src/Datadog.Trace/Debugger/Sink/Models/Diagnostics.cs#L12

Reply:
yeah  IIRC I removed it the attribute that was needed for `init` support because I was worried about introducing too much noise at once (as part of the Live Debugger open-sourcing PR), but I have nothing against trying it :) 

AFAIK that is orthogonal to `record` support though? As @shurivich mentioned, we are using records in `master` today.

Reply:
oh 😅 well, I guess they're fine then 😄 excellent!

Reply:
🤔 
![image](https://user-images.githubusercontent.com/18755388/188632654-bc42f213-bf73-45ae-8ba4-06e751861c8a.png)

Reply:
I guess we can work around it with ugliness 😅 

This triggers the warning:

```csharp
internal record AgentConfiguration(string? ConfigurationEndpoint,
                                   string? DebuggerEndpoint,
                                   string? AgentVersion,
                                   string? StatsEndpoint,
                                   string? DataStreamsMonitoringEndpoint,
                                   bool ClientDropP0s);
```

This, duplicating all the properties, and using a primary constructor, doesn't:

```csharp
internal record AgentConfiguration(string? ConfigurationEndpoint,
                                   string? DebuggerEndpoint,
                                   string? AgentVersion,
                                   string? StatsEndpoint,
                                   string? DataStreamsMonitoringEndpoint,
                                   bool ClientDropP0s)
{
    public string? ConfigurationEndpoint { get; } = ConfigurationEndpoint;

    public string? DebuggerEndpoint { get; } = DebuggerEndpoint;

    public string? AgentVersion { get; } = AgentVersion;

    public string? StatsEndpoint { get; } = StatsEndpoint;

    public string? DataStreamsMonitoringEndpoint { get; } = DataStreamsMonitoringEndpoint;

    public bool ClientDropP0s { get; } = ClientDropP0s;
}
```

Looking in sharplab, it's because the former creates properties like this:

```csharp
public string ConfigurationEndpoint
{
    [CompilerGenerated]
    get
    {
        return <ConfigurationEndpoint>k__BackingField;
    }
    [CompilerGenerated]
    init
    {
        <ConfigurationEndpoint>k__BackingField = value;
    }
}
```

whereas in the latter, they look like this:

```csharp
public string ConfigurationEndpoint
{
    [CompilerGenerated]
    get
    {
        return <ConfigurationEndpoint>k__BackingField;
    }
}
```

I guess that's still preferable to the class, but I'm still slightly sad about `IsExternalInit`!

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/3154/commits/a171c4a8f1354e9371a01a7034cfebecb9c756bd